rfcs/0014-opendal.md

```diff
@@ -0,0 +1,196 @@
+  - [Migration Plan](#migration-plan)
+
+<!-- TOC end -->
+
+Status: Work in Progress
+
+Authors:
+
+- [Li Yazhou](https://github.com/flaneur2020)
+
+<!-- TOC --><a name="background"></a>
+## Background
+
+Currently, SlateDB uses the `object_store` crate as its IO access layer. The `object_store` crate provides a minimal trait abstraction for different storage backends, which has served us well for basic storage operations.
+
+However, as development has progressed, we've encountered limitations with this minimalistic approach. Several advanced data access patterns have emerged as requirements:
+
+1. **Caching**: We implemented `CachedObjectStore` as a wrapper around `object_store::ObjectStore` to provide transparent chunked caching. This required us to maintain our own caching logic and handle cache invalidation strategies.
+
+2. **Retrying**: We implemented `RetryingObjectStore` as another wrapper layer to automatically retry operations on transient failures. This added complexity in determining which errors are retryable and managing backoff strategies.
+
+These wrapper implementations have several drawbacks:
+
+- Each feature requires implementing the full `ObjectStore` trait, leading to significant boilerplate code
+- We're duplicating efforts that are already solved problems in the broader ecosystem
+- Maintaining and testing these wrappers adds ongoing engineering overhead
+- The composition of multiple wrapper layers can be error-prone and harder to reason about
+
+There were issues that discussed about introducing some composability for `object_store` crate:
+
+- https://github.com/apache/arrow-rs-object-store/issues/14
+- https://github.com/apache/arrow-rs-object-store/issues/274
+
+Unfortunately, these issues still have little progress.
+
+We hope to have a solution that reduces the complexity of maintaining our own wrapper layers while leveraging extension capabilities from an actively developed ecosystem.
+
+**OpenDAL** offers a promising solution: it's an Apache Incubator project that provides a **composable layer system** where features like caching, retrying, and observability are production-ready layers that can be combined without custom wrapper code. This addresses our exact pain points while being actively maintained by a nice community.
+
+<!-- TOC --><a name="goals-and-constraints"></a>
+## Goals and Constraints
+
+<!-- TOC --><a name="goals"></a>
+### Goals
+
+The primary goals of migrating to OpenDAL are:
+
+1. **Reduce maintenance burden**: Eliminate the need to maintain custom wrapper layers (`CachedObjectStore`, `RetryingObjectStore`) by leveraging battle-tested ecosystem solutions
+2. **Improve extensibility**: Gain access to OpenDAL's rich layer ecosystem for future capabilities (observability, rate limiting, etc.) without additional development effort
+3. **Future-proof storage integration**: Benefit from OpenDAL's active development and broader backend support as SlateDB evolves
+
```

Comment:
Can you add a subsection in goals that explicitly lists the features we want?

- on-disk caching (and requirements)
- request racing (maybe?)
- retries
  - infinite in bg tasks
  - timeout'd retries in fg tasks
  - handles idempotent retry transparently (if file already exists and matches, return success)
- standard environment variable support (AWS_* etc)
- bugs
  - it would be nice if we didn't have to bother with #1019 (https://github.com/hyperium/hyper/issues/3995)
- etc

What else do we want from an object store library?